### PR TITLE
[Balloon] Added GET call for balloon device state

### PR DESCRIPTION
**Pull request by georgepisaltu**
## Reason for This PR

We need an API call to poll the state of the balloon device.

## Description of Changes

Added an API call to poll the state of the balloon device.

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [ ] All commits in this PR are signed (`git commit -s`).
- [ ] The reason for this PR is clearly provided (issue no. or explanation).
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] Any newly added `unsafe` code is properly documented.
- [ ] Any API changes are reflected in `firecracker/swagger.yaml`.
- [ ] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [ ] All added/changed functionality is tested.

### CODE REVIEW

File: src/devices/src/virtio/balloon/device.rs [hunk 1]
```diff
@@ -73,6 +73,15 @@ struct BalloonStat {
 // Safe because BalloonStat only contains plain data.
 unsafe impl ByteValued for BalloonStat {}
 
+// BalloonStats holds statistics returned from the stats_queue.
+#[derive(Clone, Default, Debug, PartialEq, Serialize)]
+pub struct BalloonState {
```
Comment (acatangiu) [inline src/devices/src/virtio/balloon/device.rs:78]: Suggest `BalloonConfig` or some other name; `BalloonState` misleads the reader to think of snapshots 😛

**Review by acatangiu — COMMENTED** [review]
LGTM, just needs changing `BalloonState` to `BalloonConfig` or smth similar.

File: src/api_server/swagger/firecracker.yaml [hunk 2]
```diff
@@ -60,6 +60,115 @@ paths:
...
+        default:
+          description: Internal Server Error
+          schema:
+            $ref: "#/definitions/Error"
+    patch:
+      summary: Updates a balloon device statistics polling interval.
+      description:
+        Updates an existing balloon device statistics interval, before or after machine startup.
+        Will fail if update is not possible.
+      operationId: patchStatsBalloon
```
Comment (sandreim) [inline src/api_server/swagger/firecracker.yaml:152]: This sounds like we want to update the stats. I would rename this as patchBalloonStatsInterval to make it more specific.

**Review by sandreim — COMMENTED** [review]
Changes look good. I have only one question around how the APIs would work when guest driver does not negotiate VIRTIO_BALLOON_F_STATS_VQ.

**Comment by georgepisaltu** [comment]
> Changes look good. I have only one question around how the APIs would work when guest driver does not negotiate VIRTIO_BALLOON_F_STATS_VQ.

The `GET /balloon` will be always available and will return the configuration of the device, with the flags at install time and the current target size.

The `GET /balloon/statistics` will be available only after boot and will return only the statistics, as described in the stats structure. `VIRTIO_BALLOON_F_STATS_VQ` can only be negotiated at boot time, and we only do it when the `stats_polling_interval_s` is set to a non-zero value before boot. Therefore, this request will only be available post-boot if the polling interval was set to a non-zero value pre-boot.

**Review by sandreim — DISMISSED** [review]

**Review by sandreim — APPROVED** [review]

**Review by acatangiu — APPROVED** [review]

**Pull request merged**